cpp/src/strings/search/findall.cu

```diff
@@ -21,134 +21,148 @@
-      if (prog.find<stack_size>(idx, d_str, spos, epos) <= 0) break;  // no more matches found
-      if (column_count == column_index) break;                        // found our column
-      spos = epos > spos ? epos : spos + 1;
-      epos = static_cast<int32_t>(nchars);
-      ++column_count;
+    auto const match_count = d_counts[idx];
+
+    auto d_output = d_indices[idx];
+
+    if (d_strings.is_valid(idx)) {
+      auto const d_str  = d_strings.element<string_view>(idx);
+      auto const nchars = d_str.length();
+
+      int32_t begin = 0;
+      int32_t end   = -1;
+      for (auto col_idx = 0; col_idx < match_count; ++col_idx) {
+        if (prog.find<stack_size>(idx, d_str, begin, end) > 0) {
+          auto const begin_offset = d_str.byte_offset(begin);
+          auto const end_offset   = d_str.byte_offset(end);
+          d_output[col_idx] =
+            string_index_pair{d_str.data() + begin_offset, end_offset - begin_offset};
+        }
+        begin = end + (begin == end);
+        end   = nchars;
+      }
     }
 
-    auto const result = [&] {
-      if (spos > epos) { return string_index_pair{nullptr, 0}; }
-      // convert character positions to byte positions
-      spos = d_str.byte_offset(spos);
-      epos = d_str.byte_offset(epos);
-      return string_index_pair{d_str.data() + spos, (epos - spos)};
-    }();
-
-    return result;
+    // fill the remaining entries for this row with nulls
+    thrust::fill(
+      thrust::seq, d_output.begin() + match_count, d_output.end(), string_index_pair{nullptr, 0});
   }
 };
 
 struct findall_dispatch_fn {
   reprog_device d_prog;
 
   template <int stack_size>
-  std::unique_ptr<column> operator()(column_device_view const& d_strings,
-                                     size_type column_index,
-                                     size_type const* d_find_counts,
-                                     rmm::cuda_stream_view stream,
-                                     rmm::mr::device_memory_resource* mr)
+  void operator()(column_device_view const& d_strings,
+                  size_type const* d_find_counts,
+                  indices_span& d_indices,
+                  rmm::cuda_stream_view stream)
   {
-    rmm::device_uvector<string_index_pair> indices(d_strings.size(), stream);
-
-    thrust::transform(rmm::exec_policy(stream),
-                      thrust::make_counting_iterator<size_type>(0),
-                      thrust::make_counting_iterator<size_type>(d_strings.size()),
-                      indices.begin(),
-                      findall_fn<stack_size>{d_strings, d_prog, column_index, d_find_counts});
-
-    return make_strings_column(indices.begin(), indices.end(), stream, mr);
+    thrust::for_each_n(rmm::exec_policy(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       d_strings.size(),
+                       findall_fn<stack_size>{d_strings, d_prog, d_find_counts, d_indices});
   }
 };
 }  // namespace
 
 //
```

Comment:
Is there a reason for having this empty comment line?

Reply:
Just my constant struggle against what clang-format does and how I want my code look. I preferred one extra line between the end of the namespace and the beginning of the detail function definition.